### PR TITLE
Workflow for building and pushing the TF serving image.

### PR DESCRIPTION
**Pull request by lluunn**
We want to have a workflow that automatically build and push the TF serving image from source, and run some test against it to verify it's working.
Argo is suitable for this kind of workflows, and it supports docker in docker.

In this PR, the workflows contains building+pushing part. (will add testing in followup PR).
It's using the Dockerfile in components/k8s-model-server/docker, which is a TF1.4 CPU image.

Related #50

### CODE REVIEW

**Comment by jlewi** [comment]
Please reference #50 in the PR description.

Which version of TF does this end up building a Docker image for?

File: components/k8s-model-server/testing/test-serving-container/components/workflows.libsonnet [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      // should be mounted.
+      local mountPath = "/mnt/" + "test-data-volume";
+      // testDir is the root directory for all data for a particular test run.
+      local testDir = mountPath + "/" + name;
+      // outputDir is the directory to sync to GCS to contain the output for this job.
+      local outputDir = testDir + "/output";
+      local artifactsDir = outputDir + "/artifacts";
+      local srcDir = testDir + "/src";
+      // local testing_image = "gcr.io/mlkube-testing/kubeflow-testing";
+      local testing_image = "gcr.io/kai-test2/kubeflow-testing:1.0";
```
Comment (jlewi) [inline components/k8s-model-server/testing/test-serving-container/components/workflows.libsonnet]: Why is this point at kai-cloud?
If you need to override these values make the parameters and just override them per environment.

**Comment by jlewi** [comment]
Please update the PR description. Explain the motivation for this PR and why we are using Argo.

File: components/k8s-model-server/testing/test-serving-container/components/params.libsonnet [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  components: {
+    // Component-level parameters, defined initially from 'ks prototype use ...'
+    // Each object below should correspond to a component in the components/ directory
+    workflows: {
+      bucket: "mlkube-testing_temp",
+      commit: "master",
+      name: "new4",
+      namespace: "kubeflow-test-infra",
+      prow_env: "REPO_OWNER=google,REPO_NAME=kubeflow,PULL_BASE_SHA=master",
```
[the remark this answers is not in the document]
Comment (lluunn) [reply components/k8s-model-server/testing/test-serving-container/components/params.libsonnet]: done

File: components/k8s-model-server/testing/test-serving-container/components/workflows.libsonnet (hunk 1 above)
Comment (lluunn) [reply components/k8s-model-server/testing/test-serving-container/components/workflows.libsonnet]: I made some change to the Dockerfile (install docker).
Made it a param now.

File: components/k8s-model-server/testing/test-serving-container/debug_pod.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# This pod is useful for starting a shell that you can use to interactively debug our tests
```
[the remark this answers is not in the document]
Comment (lluunn) [reply components/k8s-model-server/testing/test-serving-container/debug_pod.yaml]: done

**Comment by lluunn** [comment]
Thanks, PTAL

**Comment by jlewi** [comment]
Lets change the directory to 

```
testing/test-serving-container  -> images/releaser
```

File: components/k8s-model-server/testing/test-serving-container/components/workflows.jsonnet [hunk 4]
```diff
@@ -0,0 +1,15 @@
+local params = std.extVar("__ksonnet/params").components.workflows;
+
+local k = import 'k.libsonnet';
+local workflows = import 'workflows.libsonnet';
+local namespace = params.namespace;
+local serving_image = params.serving_image;
+local testing_image = params.testing_image;
+
+// TODO(jlewi): Can we make name default so some random unique value?
```
Comment (jlewi) [inline components/k8s-model-server/testing/test-serving-container/components/workflows.jsonnet]: Delete this TODO please.

File: components/k8s-model-server/testing/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Building and testing the TF serving image.
+
+work in progress.
```
Comment (jlewi) [inline components/k8s-model-server/testing/README.md]: The README should explain how I can trigger this workflow to build a TF Serving image. The README should explain the following

    1. How to set which version of the Kubeflow source repository is used
    2. How to set the GCR registry where the image is published to
    3. Which version of TF serving is used.

**Comment by lluunn** [comment]
PTAL

File: components/k8s-model-server/images/README.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+```
+
+## Checkout
+The version of kubeflow is determined by the parameters such as `PULL_NUMBER` or `PULL_BASE_SHA`.
+See [checkout.sh](https://github.com/kubeflow/testing/blob/master/images/checkout.sh) for more detail.
+
+## Build and Push
+Set the gcr registry to be published to by `ks set param workflows serving_image gcr.io/xxx/xxx:xx`.
+
+The TF serving image that the workflow is currently using is TF1.4 CPU, located at `components/k8s-model-server/docker`.
```
Comment (jlewi) [inline components/k8s-model-server/images/README.md]: The text seems wrong. I assume you mean "TF version" not TF serving image?

File: components/k8s-model-server/images/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+The workflow will checkout the source code from kubeflow/kubeflow, then build and push the image.
+
+Deploy the workflow by:
+```
+ks apply prow -c workflows
+```
+
+## Checkout
+The version of kubeflow is determined by the parameters such as `PULL_NUMBER` or `PULL_BASE_SHA`.
+See [checkout.sh](https://github.com/kubeflow/testing/blob/master/images/checkout.sh) for more detail.
```
Comment (jlewi) [inline components/k8s-model-server/images/README.md]: Add a link to this list here:
https://github.com/kubernetes/test-infra/tree/master/prow#job-environment-variables

**Comment by lluunn** [comment]
done

**Pull request merged**